Sources/WalletConnect/Engine/PairingEngine.swift

```
@@ -1,6 +1,73 @@
+import Foundation
+
 final class PairingEngine: SequenceEngine {
     
-    func respond(to proposal: SequenceProposal, completion: @escaping (Result<String, Error>) -> Void) {
-        // TODO: Respond to pairing proposal
+    private let pendingPairings: SequenceSubscribing
+    private let settledPairings: SequenceSubscribing!
+    
+    private let relayer: Relaying
+    private let crypto: Crypto
+    
+    init(relay: Relaying, crypto: Crypto) {
+        self.relayer = relay
+        self.crypto = crypto
+        self.pendingPairings = Subscription(relay: relay)
+        self.settledPairings = Subscription(relay: relay)
+    }
+    
+    func respond(to proposal: PairingType.Proposal, completion: @escaping (Result<String, Error>) -> Void) {
+        let privateKey = Crypto.X25519.generatePrivateKey()
+        let selfPublicKey = privateKey.publicKey.toHexString()
+        
+        let pendingPairing = PairingType.Pending(
+            status: .responded,
+            topic: proposal.topic,
+            relay: proposal.relay,
+            self: PairingType.Participant(publicKey: selfPublicKey),
+            proposal: proposal)
+        
+        pendingPairings.set(topic: proposal.topic, sequenceData: .pending(pendingPairing))
+        
+        // settle on topic B
+        let agreementKeys = try! Crypto.X25519.generateAgreementKeys(
+            peerPublicKey: Data(hex: proposal.proposer.publicKey),
+            privateKey: privateKey)
+        let topicB = agreementKeys.sharedSecret.sha256().toHexString()
+        
+        let controllerKey = proposal.proposer.controller ? proposal.proposer.publicKey : selfPublicKey
```

for now Swift client is always a controller, so maybe we should throw an error when dapp wants to set itself as a controller as it could be a malicious action 🤔

---

I want to understand better the controller actions, but you're right to be concerned about it, we might need to handle this